File: backend/src/models/fineModel.js
const { FINE_STATUS } = require('../config/constants');

class Fine {
  constructor(
    fineId,
    userId,
    reason,
    amount,
    status = FINE_STATUS.PAYMENT_PENDING,
    createdBy
  ) {
    this.fineId = fineId;
    this.userId = userId;
    this.reason = reason;
    this.amount = Number(amount);
    this.status = status;
    this.createdBy = createdBy; // admin who created the fine
    this.createdAt = new Date().toISOString();
    this.updatedAt = new Date().toISOString();
  }

  static fromFirestore(doc) {
    const data = doc.data();
    const fine = new Fine(
      doc.id,
      data.userId,
      data.reason,
      data.amount,
      data.status,
      data.createdBy
    );
    fine.createdAt = data.createdAt;
    fine.updatedAt = data.updatedAt;
    return fine;
  }

  toFirestore() {
    return {
      fineId: this.fineId,
      userId: this.userId,
      reason: this.reason,
      amount: this.amount,
      status: this.status,
      createdBy: this.createdBy,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }
}

module.exports = Fine;

File: backend/frontend/lib/models/fine_model.dart
import 'package:post_app/models/enums.dart';

class FineModel {
  final String fineId;
  final String userId;
  final String reason;
  final double amount;
  final FineStatus status;
  final String createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  FineModel({
    required this.fineId,
    required this.userId,
    required this.reason,
    required this.amount,
    required this.status,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
  });

  factory FineModel.fromJson(Map<String, dynamic> json) {
    return FineModel(
      fineId: json['fineId'] as String,
      userId: json['userId'] as String,
      reason: json['reason'] as String,
      amount: (json['amount'] as num).toDouble(),
      status: FineStatus.fromJson(json['status'] as String),
      createdBy: json['createdBy'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'fineId': fineId,
      'userId': userId,
      'reason': reason,
      'amount': amount,
      'status': status.toJson(),
      'createdBy': createdBy,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

// --- Request/Response Models for Fine ---

class CreateFineRequest {
  final String userId;
  final String reason;
  final double amount;

  CreateFineRequest({
    required this.userId,
    required this.reason,
    required this.amount,
  });

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'reason': reason,
      'amount': amount,
    };
  }
}

class AdminUpdateFineStatusRequest {
  final FineStatus status;

  AdminUpdateFineStatusRequest({required this.status});

  Map<String, dynamic> toJson() {
    return {
      'status': status.toJson(),
    };
  }
}

class CreateFineResponse {
  final String message;
  final FineModel fine;

  CreateFineResponse({
    required this.message,
    required this.fine,
  });

  factory CreateFineResponse.fromJson(Map<String, dynamic> json) {
    return CreateFineResponse(
      message: json['message'] as String,
      fine: FineModel.fromJson(json['fine'] as Map<String, dynamic>),
    );
  }
}